### PR TITLE
Look up documents before retrieving, improve id/key naming conventions

### LINKED ISSUE #394: GET /api/v1/advisory/urn:uuid:abcd/download does not work
Response from the backend 

```json
{
    "error": "Unsupported hash algorithm",
    "message": "Unsupported algorithm urn:uuid"
}
```

### PR DESCRIPTION
**Pull request by ctron**
Fix #394: Before downloading the actual content, we look up the sha256 digest from the database and use that to fetch the content from the storage. This is done for advisories, but also for SBOMs.

Also:
  * When uploading a document, return the `Id` of the database instead of some ID field from the document
  * Rename `HashOrUuidKey` to `Id`
  * Make it clearer what key/id means for the storage (`StorageKey`)

### CODE REVIEW

**Comment by ctron** [comment]
There's one catch. It relies on `fetch_advisory`, which is also limited and allows only retrieving by `Uuid` and `Sha256`. But I guess that another problem to fix.

**Review by bobmcwhirter — APPROVED** [review]

**Comment by ctron** [comment]
The CI disagrees. I'll fix that tomorrow.

**Review by carlosthe19916 — APPROVED** [review]

**Comment by carlosthe19916** [comment]
Tested and the REST API worked as expected, thanks!

File: modules/fundamental/src/advisory/endpoints/mod.rs
Comment (jcrossley3) [inline modules/fundamental/src/advisory/endpoints/mod.rs:155]: minor style nit. might warrant adding a `find_hash` method to `AdvisoryHead`

```suggestion
    let id = Id::from_str(&key).map_err(Error::HashKey)?;
    let Some(hash_key) = advisory.fetch_advisory(id, ()).await?.and_then(|adv| {
        adv.head
            .hashes
            .into_iter()
            .find(|h| matches!(h, Id::Sha256(_)))
    }) else {
        return Ok(HttpResponse::NotFound().finish());
    };
```

**Review by jcrossley3 — CHANGES_REQUESTED** [review]

**Comment by ctron** [comment]
I updated the PR and would value a second review.

The same issue was present for SBOMs. That should not also be fixed.

I also did refactor the storage part in a way that it is clear that there is a "storage key", which is opaque on the API. And there's a translation from the `Id` (and `Vec<Id>`). Which I is a better approach to the `find_sha256` function.

File: modules/fundamental/src/advisory/endpoints/test.rs
Comment (jcrossley3) [inline modules/fundamental/src/advisory/endpoints/test.rs:422]: As a user, I took some comfort seeing the identifier of the doc I posted returned upon success. The opaque uuid gives me less comfort. Not for this PR, but do you think it's worth returning a struct with `uuid` and `identifier` fields upon success?

File: modules/fundamental/src/advisory/endpoints/mod.rs
Comment (jcrossley3) [inline modules/fundamental/src/advisory/endpoints/mod.rs:154]: much cleaner!

File: modules/storage/src/service/fs.rs
Comment (jcrossley3) [inline modules/storage/src/service/fs.rs:98]: I don't quite understand why storage needs to calculate all these for every write. It's only gonna use one of them, correct?

**Review by jcrossley3 — COMMENTED** [review]
I like the `StorageKey`

File: modules/storage/src/service/fs.rs
Comment (ctron) [reply modules/storage/src/service/fs.rs:98]: Correct. But it returns them after the insert. My idea was to refactor in another step, passing those into the loaders, so that they might be able to use that info. Right now, they do this during the read. Which also works.

**Comment by ctron** [comment]
@jcrossley3 unless you think the PR is getting too big, I could amend it tomorrow, adding the struct and trying to pass through the list of hashes.

**Comment by jcrossley3** [comment]
> @jcrossley3 unless you think the PR is getting too big, I could amend it tomorrow, adding the struct and trying to pass through the list of hashes.

I do think it's getting pretty big. Can you change the title to be more descriptive and move the "Fix #394" into the description before merging?

**Comment by ctron** [comment]
I continued work in https://github.com/trustification/trustify/pull/405

**Pull request merged**
